File: iq/23/substring/find.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package substring

import (
	"iter"
	"unicode/utf8"
)

func Find(s string, n int) string {
	var longest []byte
	for buf := range substrings(s, n) {
		if len(buf) > len(longest) {
			longest = buf
		}
	}
	return string(longest)
}

func substrings(s string, n int) iter.Seq[[]byte] {
	return func(yield func([]byte) bool) {
		w := newWindow(s, n)
		for !w.end() {
			w.expand()
			if !yield(w.bytes()) {
				break
			}
			w.shrink()
		}
	}
}

type window struct {
	buf   []byte
	nrepl int

	freq    map[rune]int
	freqMax int

	left  int
	right int
	count int
}

func newWindow(s string, n int) *window {
	return &window{
		buf:   []byte(s),
		nrepl: n,
		freq:  make(map[rune]int),
	}
}

func (w *window) end() bool {
	return w.right == len(w.buf)
}

func (w *window) bytes() []byte {
	return w.buf[w.left:w.right]
}

func (w *window) expand() {
	for w.right < len(w.buf) {
		r, rs := utf8.DecodeRune(w.buf[w.right:])
		if r == utf8.RuneError {
			if rs == 1 {
				// encoding error
				break
			}
			// empty
			break
		}
		if ok := w.expandOne(r, rs); !ok {
			break
		}
	}
}

func (w *window) expandOne(r rune, rs int) bool {
	canExpand := func(r rune) bool {
		rfreq := w.freq[r] + 1
		count := w.count + 1
		nmax := w.freqMax
		if rfreq > nmax {
			nmax = rfreq
		}
		replacements := count - nmax
		return replacements <= w.nrepl
	}
	if !canExpand(r) {
		return false
	}
	w.freq[r] += 1
	w.right += rs
	w.count += 1
	if n := w.freq[r]; n > w.freqMax {
		w.freqMax = n
	}
	return true
}

func (w *window) shrink() {
	r, rs := utf8.DecodeRune(w.buf[w.left:])
	if r == utf8.RuneError {
		if rs == 1 {
			// encoding error
			return
		}
		// empty
		return
	}
	w.freq[r] -= 1
	w.left += rs
	w.count -= 1
}

File: iq/23/substring/find.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "iq/23/substring/find.h"

#include <algorithm>
#include <cstddef>
#include <string_view>
#include <unordered_map>

namespace iq::substring {

std::string_view Find(std::string_view s, std::size_t n) {
  std::string_view substr;
  std::unordered_map<char, std::size_t> freq;
  std::size_t freq_max = 0;  // maximum frequency of the last valid window
  std::size_t start = 0;
  std::size_t end = 0;
  while (end < s.size()) {
    const char ch = s[end];
    end += 1;
    freq[ch] += 1;
    freq_max = std::max(freq_max, freq[ch]);
    if (const std::size_t replacements = (end - start) - freq_max;
        replacements > n) {
      const char ch = s[start];
      freq[ch] -= 1;
      start += 1;
    }
    if (const std::size_t size = end - start; size > substr.size()) {
      substr = s.substr(start, size);
    }
  }
  return substr;
}

}  // namespace iq::substring
